Firmware SDK: add device authentication section (#283)
golioth/docs: docs/firmware/golioth-firmware-sdk/authentication/certificate-auth.md

* Create device certificates signed using the private root key
* Upload the public root certificate to Golioth
* Provide the device certificate and key to your device

:::note Safeguard your private root key

You retain the **private** root key, it is not shared with Golioth. Only you are
able to sign new device certificates.

:::

## Generate Certificates

### Create a self-signed root certificate

For testing we will use `openssl` to generate a self-signed root certificate.

```shell
SERVER_NAME='golioth'

# Generate an elliptic curve private key 
# Run `openssl ecparam -list_curves` to list all available algorithms
# Keep this key safe! Anyone who has it can sign authentic-looking device certificates
openssl ecparam -name prime256v1 -genkey -noout -out "${SERVER_NAME}.key.pem"

# Create and self-sign a corresponding public key / certificate
openssl req -x509 -new -nodes \
    -key "${SERVER_NAME}.key.pem" \
    -sha256 -subj "/C=BR/CN=Root ${SERVER_NAME}" \
    -days 1024 -out "${SERVER_NAME}.crt.pem"
```

This will generate the following files:

* **Signed root certificate:** `golioth.crt.pem`
* **Private root key:** `golioth.key.pem`

### Create and sign device certificates

> Now generate and sign a device certificate. The resulting key/csr will be

Reviewer: The CSR (certificate signing request) stays in PEM format. The signed certificate and the private key are in DER
```suggestion
Now generate and sign a device certificate. The resulting key/cert will be
```

converted to the `DER` binary format for use with the Golioth Zephyr SDK.

```shell
PROJECT_SLUG='your-golioth-projectID'
PRIMARY_HARDWARE_ID='choose-hardware-id-for-this-device'
SERVER_NAME='golioth'

> CLIENT_NAME="${PROJECT_SLUG}-${PRIMARY_HARDWARE_ID}"

Reviewer: The CN in the certificate actually maps to the device name, not the Primary Hardware ID. See https://www.notion.so/golioth/Identifying-Devices-From-Their-Certificates-93074ae80ed34b80a2f0d33f98abf269
```suggestion
CLIENT_NAME="${PROJECT_SLUG}-${DEVICE_NAME}"
```

# Generate an elliptic curve private key
openssl ecparam -name prime256v1 -genkey -noout -out "${CLIENT_NAME}.key.pem"

# Create a certificate signing request (CSR)
# (this is what you would normally give to your CA / PKI to sign)
openssl req -new \
    -key "${CLIENT_NAME}.key.pem" \

>     -subj "/C=BR/O=${PROJECT_SLUG}/CN=${PRIMARY_HARDWARE_ID}" \

Reviewer: ```suggestion
    -subj "/C=BR/O=${PROJECT_SLUG}/CN=${DEVICE_NAME}" \
```

    -out "${CLIENT_NAME}.csr.pem"

# Sign the certificate (CSR) using the previously generated self-signed root certificate
openssl x509 -req \
    -in "${CLIENT_NAME}.csr.pem" \
    -CA "${SERVER_NAME}.crt.pem" \
    -CAkey "${SERVER_NAME}.key.pem" \
    -CAcreateserial \
    -out "${CLIENT_NAME}.crt.pem" \
    -days 500 -sha256

# Convert device certificates to DER format
openssl x509 -in ${CLIENT_NAME}.crt.pem -outform DER -out ${CLIENT_NAME}.crt.der
openssl ec -in ${CLIENT_NAME}.key.pem -outform DER -out ${CLIENT_NAME}.key.der
```

This will generate the following files:

> * **Signed device certificate:** `<projectID>-<primary-hardware-ID].crt.pem`

Reviewer: `s/primary-hardware-ID/device-name` for these five lines

certificates in [the Zephyr-based `certificate_provisioning`
sample](https://github.com/golioth/golioth-firmware-sdk/tree/main/examples/zephyr/certificate_provisioning).

### Upload public root key

Upload the signed root certificate (the public key) to your Golioth project. This
will be used to authenticate each device.

1. Log into the [Golioth Web Console](https://console.golioth.io) and select
   your project
2. Navigate to [`Project Settings` &rarr;
   `Certificates`](https://console.golioth.io/project-settings/certificates)
3. Click `Add a Client Certificate Authority` and upload your `golioth.crt.pem`
   file.

![Uploading the public root certificate](../../assets/cert-auth-upload-public-key.jpg)

:::caution

Do not upload the `golioth.key.pem` file. This is your private key used to sign
new device credentials and you are the only one who should have access to it.

:::

## Connect your device

With the public root certificate uploaded to Golioth, and the device
certificate/key in use in your firmware, your device will automatically connect
and authenticate to your project. Look in the [Golioth Web
Console](https://console.golioth.io) to verify a device matching the

> `PRIMARY_HARDWARE_ID` you supplied when creating device credentials is now

Reviewer: ```suggestion
`DEVICE_NAME` you supplied when creating device credentials is now
```
